src/main/java/cuchaz/enigma/network/EnigmaServer.java

```diff
@@ -108,21 +113,25 @@ private void acceptClient() throws IOException {
 		thread.start();
 	}
 
-	public synchronized void stop() {
-		if (!socket.isClosed()) {
-			for (Socket client : clients) {
-				kick(client, "disconnect.server_closed");
-			}
-			try {
-				socket.close();
-			} catch (IOException e) {
-				System.err.println("Failed to close server socket");
-				e.printStackTrace();
+	public void stop() {
+		runOnThread(() -> {
+			if (!socket.isClosed()) {
+				for (Socket client : clients) {
+					kick(client, "disconnect.server_closed");
+				}
+				try {
+					socket.close();
+				} catch (IOException e) {
+					System.err.println("Failed to close server socket");
+					e.printStackTrace();
+				}
 			}
-		}
+		});
 	}
 
 	public void kick(Socket client, String reason) {
+		if (!clients.contains(client)) return;
+
 		sendPacket(client, new KickS2CPacket(reason));
 
 		clients.remove(client);
```

Comment:
This could be merged with the `clients.contains` check above

Reply:
Which check? There's no occurrence of `clients.contains` above that line

Reply:
`if (!clients.remove(client)) return;`
`remove` returns whether it actually removed something.

Reply:
Ahh, right.